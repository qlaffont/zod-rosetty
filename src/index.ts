import type { RosettyReturn } from 'rosetty';
import { defaultErrorMap, ZodErrorMap, ZodIssueCode, ZodParsedType } from 'zod';

export type ZodLocaleMap = {
  zod: {
    errors: {
      invalid_type: string;
      invalid_type_received_undefined: string;
      invalid_literal: string;
      unrecognized_keys: string;
      invalid_union: string;
      invalid_union_discriminator: string;
      invalid_enum_value: string;
      invalid_arguments: string;
      invalid_return_type: string;
      invalid_date: string;
      custom: string;
      invalid_intersection_types: string;
      not_multiple_of: string;
      not_finite: string;
      invalid_string: {
        email: string;
        url: string;
        uuid: string;
        cuid: string;
        regex: string;
        datetime: string;
        startsWith: string;
        endsWith: string;
      };
      too_small: {
        array: {
          exact: string;
          inclusive: string;
          not_inclusive: string;
        };
        string: {
          exact: string;
          inclusive: string;
          not_inclusive: string;
        };
        number: {
          exact: string;
          inclusive: string;
          not_inclusive: string;
        };
        set: {
          exact: string;
          inclusive: string;
          not_inclusive: string;
        };
        date: {
          exact: string;
          inclusive: string;
          not_inclusive: string;
        };
      };
      too_big: {
        array: {
          exact: string;
          inclusive: string;
          not_inclusive: string;
        };
        string: {
          exact: string;
          inclusive: string;
          not_inclusive: string;
        };
        number: {
          exact: string;
          inclusive: string;
          not_inclusive: string;
        };
        set: {
          exact: string;
          inclusive: string;
          not_inclusive: string;
        };
        date: {
          exact: string;
          inclusive: string;
          not_inclusive: string;
        };
      };
    };
    validations: {
      email: string;
      url: string;
      uuid: string;
      cuid: string;
      regex: string;
      datetime: string;
    };
    types: {
      function: string;
      number: string;
      string: string;
      nan: string;
      integer: string;
      float: string;
      boolean: string;
      date: string;
      bigint: string;
      undefined: string;
      symbol: string;
      null: string;
      array: string;
      object: string;
      unknown: string;
      promise: string;
      void: string;
      never: string;
      map: string;
      set: string;
    };
  };
};

const jsonStringifyReplacer = (_: string, value: unknown): unknown => {
  if (typeof value === 'bigint') {
    return value.toString();
  }
  return value;
};

function joinValues<T extends unknown[]>(array: T, separator = ' | '): string {
  return array
    .map((val) => (typeof val === 'string' ? `'${val}'` : val))
    .join(separator);
}

export const zodRosettyMap: ZodErrorMap = (issue, ctx) => {
  let message: string;
  message = defaultErrorMap(issue, ctx).message;

  switch (issue.code) {
    case ZodIssueCode.invalid_type:
      if (issue.received === ZodParsedType.undefined) {
        message = `zod.errors.invalid_type_received_undefined`;
      } else {
        message = `zod.errors.invalid_type - ${JSON.stringify({
          expected: `zod.types.${issue.expected}`,
          received: `zod.types.${issue.received}`,
        })}`;
      }
      break;
    case ZodIssueCode.invalid_literal:
      message = `zod.errors.invalid_literal - ${JSON.stringify({
        expected: JSON.stringify(issue.expected, jsonStringifyReplacer),
      })}`;
      break;
    case ZodIssueCode.unrecognized_keys:
      message = `zod.errors.unrecognized_keys - ${JSON.stringify({
        keys: joinValues(issue.keys, ', '),
        count: issue.keys.length,
      })}`;
      break;
    case ZodIssueCode.invalid_union:
      message = 'zod.errors.invalid_union';
      break;
    case ZodIssueCode.invalid_union_discriminator:
      message = `zod.errors.invalid_union_discriminator - ${JSON.stringify({
        options: joinValues(issue.options),
      })}`;
      break;
    case ZodIssueCode.invalid_enum_value:
      message = `zod.errors.invalid_enum_value - ${JSON.stringify({
        options: joinValues(issue.options),
        received: issue.received,
      })}`;
      break;
    case ZodIssueCode.invalid_arguments:
      message = 'zod.errors.invalid_arguments';
      break;
    case ZodIssueCode.invalid_return_type:
      message = 'zod.errors.invalid_return_type';
      break;
    case ZodIssueCode.invalid_date:
      message = 'zod.errors.invalid_date';
      break;
    case ZodIssueCode.invalid_string:
      if (typeof issue.validation === 'object') {
        if ('startsWith' in issue.validation) {
          message = `zod.errors.invalid_string.startsWith - ${JSON.stringify({
            startsWith: issue.validation.startsWith,
          })}`;
        } else if ('endsWith' in issue.validation) {
          message = `zod.errors.invalid_string.endsWith - ${JSON.stringify({
            endsWith: issue.validation.endsWith,
          })}`;
        }
      } else {
        message = `zod.errors.invalid_string.${
          issue.validation
        } - ${JSON.stringify({
          validation: `zod.validations.${issue.validation}`,
        })}`;
      }
      break;
    case ZodIssueCode.too_small:
      // eslint-disable-next-line no-case-declarations
      const minimum =
        issue.type === 'date'
          ? new Date(issue.minimum as number)
          : issue.minimum;
      message = `zod.errors.too_small.${issue.type}.${
        issue.exact ? 'exact' : issue.inclusive ? 'inclusive' : 'not_inclusive'
      } - ${JSON.stringify({
        minimum,
        count: typeof minimum === 'number' ? minimum : undefined,
      })}`;
      break;
    case ZodIssueCode.too_big:
      // eslint-disable-next-line no-case-declarations
      const maximum =
        issue.type === 'date'
          ? new Date(issue.maximum as number)
          : issue.maximum;
      message = `zod.errors.too_big.${issue.type}.${
        issue.exact ? 'exact' : issue.inclusive ? 'inclusive' : 'not_inclusive'
      } - ${JSON.stringify({
        maximum,
        count: typeof maximum === 'number' ? maximum : undefined,
      })}`;
      break;
    case ZodIssueCode.custom:
      message = 'errors.custom';
      break;
    case ZodIssueCode.invalid_intersection_types:
      message = 'zod.errors.invalid_intersection_types';
      break;
    case ZodIssueCode.not_multiple_of:
      message = `zod.errors.not_multiple_of - ${JSON.stringify({
        multipleOf: issue.multipleOf,
      })}`;
      break;
    case ZodIssueCode.not_finite:
      message = 'zod.errors.not_finite';
      break;
    default:
  }

  return { message };
};

export const translateZodErrorMessage = (
  error: { message: string | ZodErrorMap } | undefined,
  t: RosettyReturn<ZodLocaleMap>['t'],
) => {
  if (!error || !t) {
    return undefined;
  }

  const message =
    typeof error.message === 'string' && !error.message.startsWith('[')
      ? error.message
      : //@ts-ignore
        JSON.parse(error.message)[0].message;

  const [key, valuesString] = message.split('-').map((i: string) => i.trim());

  if (!valuesString) {
    //@ts-ignore
    return t(key);
  }
  const values = JSON.parse(valuesString);

  for (const [k, v] of Object.entries(values)) {
    if (typeof v === 'string' && v.startsWith('zod.')) {
      const [key, valuesString] = v.split('-').map((i) => i.trim());
      //@ts-ignore
      values[k] = t(key, valuesString ? JSON.parse(valuesString) : {});
      continue;
    }
  }

  //@ts-ignore
  return t(key, values);
};
