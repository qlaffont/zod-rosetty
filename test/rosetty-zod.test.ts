/**
 * @jest-environment node
 */
/* eslint-disable @typescript-eslint/no-explicit-any */

import { rosetty } from 'rosetty';
import { z, ZodError, ZodErrorMap } from 'zod';

import { translateZodErrorMessage, zodRosettyMap } from '../src';

z.setErrorMap(zodRosettyMap);

const rosettyMap: Partial<ZodErrorMap> = {
  zod: {
    errors: {
      invalid_type_received_undefined: 'Required',
      invalid_string: {
        email: 'Invalid {{validation}}',
      },
      invalid_type: 'Expected {{expected}}, received {{received}}',
    },
    validations: {
      email: 'email',
    },
    types: {
      string: 'string',
      number: 'number',
    },
  },
};

const getMessageError = (zodError: ZodError) => {
  return JSON.parse(zodError.message)[0].message;
};

describe('Rosetty Zod', () => {
  describe('makeZodRosettyMap', () => {
    it('should be able to convert errors with values to string', async () => {
      const schema = z.string().email();
      //@ts-ignore
      expect(getMessageError(schema.safeParse('test').error)).toBe(
        'zod.errors.invalid_string.email - {"validation":"zod.validations.email"}',
      );
    });

    it('should be able to convert errors ', async () => {
      const schema = z.string();
      //@ts-ignore
      expect(getMessageError(schema.safeParse().error)).toBe(
        'zod.errors.invalid_type_received_undefined',
      );
    });
  });

  describe('translateZodErrorMessage', () => {
    const { t } = rosetty({ en: { dict: rosettyMap, locale: "en-GB" } }, 'en');

    it('should be able to convert errors with values to string', async () => {
      const schema = z.string().email();
      //@ts-ignore
      expect(translateZodErrorMessage(schema.safeParse('test').error, t)).toBe(
        'Invalid email',
      );
      //@ts-ignore
      expect(translateZodErrorMessage(schema.safeParse(123).error, t)).toBe(
        'Expected string, received number',
      );
    });

    it('should be able to convert errors ', async () => {
      const schema = z.string();
      //@ts-ignore
      expect(translateZodErrorMessage(schema.safeParse().error, t)).toBe(
        'Required',
      );
    });

    it('should return nothing if no translation function ', async () => {
      const schema = z.string();
      //@ts-ignore
      expect(translateZodErrorMessage(schema.safeParse().error)).toBe(
        undefined,
      );
    });
  });

  describe('test all schema', () => {
    it('should be able to validate all schema', () => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      let schema: any = z.string();
      expect(
        getMessageError(schema.safeParse(undefined).error),
      ).toMatchSnapshot();

      schema = z.literal('CHECK');

      expect(getMessageError(schema.safeParse('tr').error)).toMatchSnapshot();

      schema = z.object({}).strict();

      expect(
        getMessageError(schema.safeParse({ tot: true }).error),
      ).toMatchSnapshot();

      schema = z.union([
        z.object({ a: z.number() }),
        z.object({ b: z.number() }),
      ]);

      expect(
        getMessageError(schema.safeParse({ tot: true }).error),
      ).toMatchSnapshot();

      schema = z.enum(['Foo', 'Bar']);

      expect(getMessageError(schema.safeParse('to').error)).toMatchSnapshot();

      schema = z.string().startsWith('toto');

      expect(getMessageError(schema.safeParse('qwe').error)).toMatchSnapshot();

      schema = z.string().endsWith('toto');

      expect(getMessageError(schema.safeParse('qwe').error)).toMatchSnapshot();

      schema = z.string().min(5);

      expect(getMessageError(schema.safeParse('qwe').error)).toMatchSnapshot();

      schema = z.string().max(2);

      expect(getMessageError(schema.safeParse('qwe').error)).toMatchSnapshot();

      schema = z.custom((v) => v === true);

      expect(getMessageError(schema.safeParse(false).error)).toMatchSnapshot();

      schema = z.intersection(
        z.number(),
        z.number().transform((x) => x + 1),
      );

      expect(getMessageError(schema.safeParse(123).error)).toMatchSnapshot();

      schema = z.number().multipleOf(3);

      expect(getMessageError(schema.safeParse(2).error)).toMatchSnapshot();

      schema = z.number().finite();

      expect(
        getMessageError(schema.safeParse(Number.POSITIVE_INFINITY).error),
      ).toMatchSnapshot();

      schema = z.number().finite();

      expect(
        getMessageError(schema.safeParse(BigInt(3)).error),
      ).toMatchSnapshot();

      schema = z.discriminatedUnion('type', [
        z.object({ type: z.literal('a'), a: z.string() }),
        z.object({ type: z.literal('b'), b: z.string() }),
      ]);

      expect(
        getMessageError(schema.safeParse({ type: 'x', a: 'abc' }).error),
      ).toMatchSnapshot();

      schema = z.date();

      expect(
        getMessageError(schema.safeParse(new Date('invalid')).error),
      ).toMatchSnapshot();

      schema = z
        .function(z.tuple([z.string()]), z.boolean())
        .implement((arg) => {
          return arg.length as any;
        });

      try {
        schema('12' as any);
      } catch (err) {
        //@ts-ignore
        expect(getMessageError(err)).toMatchSnapshot();
      }

      schema = z
        .function(z.tuple([z.string()]), z.boolean())
        .implement((arg) => {
          return arg.length as any;
        });

      try {
        schema(12 as any);
      } catch (err) {
        //@ts-ignore
        expect(getMessageError(err)).toMatchSnapshot();
      }
    });
  });
});
