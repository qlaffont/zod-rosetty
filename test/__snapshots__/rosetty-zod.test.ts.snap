// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rosetty Zod test all schema should be able to validate all schema 1`] = `"zod.errors.invalid_type_received_undefined"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 2`] = `"zod.errors.invalid_literal - {"expected":"\\"CHECK\\""}"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 3`] = `"zod.errors.unrecognized_keys - {"keys":"'tot'","count":1}"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 4`] = `"zod.errors.invalid_union"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 5`] = `"zod.errors.invalid_enum_value - {"options":"'Foo' | 'Bar'","received":"to"}"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 6`] = `"zod.errors.invalid_string.startsWith - {"startsWith":"toto"}"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 7`] = `"zod.errors.invalid_string.endsWith - {"endsWith":"toto"}"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 8`] = `"zod.errors.too_small.string.inclusive - {"minimum":5,"count":5}"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 9`] = `"zod.errors.too_big.string.inclusive - {"maximum":2,"count":2}"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 10`] = `"errors.custom"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 11`] = `"zod.errors.invalid_intersection_types"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 12`] = `"zod.errors.not_multiple_of - {"multipleOf":3}"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 13`] = `"zod.errors.not_finite"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 14`] = `"zod.errors.invalid_type - {"expected":"zod.types.number","received":"zod.types.bigint"}"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 15`] = `"zod.errors.invalid_union_discriminator - {"options":"'a' | 'b'"}"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 16`] = `"zod.errors.invalid_date"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 17`] = `"zod.errors.invalid_return_type"`;

exports[`Rosetty Zod test all schema should be able to validate all schema 18`] = `"zod.errors.invalid_arguments"`;
